### PR TITLE
Support sampling behavior when composite filter executing action

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
Support sampling behavior after xDS matching returns positive. 
A dice roll is performed after the match, and the action will only be taken if the dice roll returns true.

This is to address issue: 
https://github.com/envoyproxy/envoy/issues/30147

The API design is discussed under:
https://github.com/envoyproxy/envoy/pull/34325

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanjunxiang-google) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:68]: The reason to name this as skip_percent, not sampling_percent is for backward compatibility. i.e, by default with this field not configured, zero percent will be skipped, i.e,, 100% traffic will be processed by the action filter.

Comment (htuch) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto:68]: I don't think we need to do that. Since this is a message field, we can use the absence of the field to enable 100% of traffic sampled, and then when set it will indicate the sample rate.

Comment (htuch) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto:68]: I'm good with this API for the various reasons already articulated in discussions. I know @markdroth had some lingering concerns so I'll defer to him for LGTM.

Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto]: Thanks for the comments! Changed the proto field name to be "sample_percent".

**Review by rshriram — APPROVED** [review]

**Review by htuch — COMMENTED** [review]
/lgtm api

**Review by tyxia — COMMENTED** [review]
LGTM overall.

Thanks for driving and pushing the development of this feature forward!

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc:121]: Sampling has to be done in dateplane. I.e, it does coin toss every time it receives the request.  It should not be done at the control plane during configuration parsing.

File: source/extensions/filters/http/composite/action.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.h:82]: Same as above. Only the call back, i.e, the returned lamda function is executed in data plane when handling request. This function itself is executed during control plane during configuration parsing.

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc:35]: PANIC_ON_PROTO_ENUM_SENTINEL_VALUES is defined as below:  https://github.com/envoyproxy/envoy/blob/de6b730c3f59e8ee3ee756c60bc826d7ede295e1/source/common/common/assert.h#L309

It can not capture a bad configuration, i.e, if denominator is set to a number , like 5.

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc]: done

File: source/extensions/filters/http/composite/action.cc
Comment (tyxia) [reply source/extensions/filters/http/composite/action.cc:35]: Catching invalid configuration like 5 should be PGV's responsibility, as `[(validate.rules).enum = {defined_only: true}];` is defined there.

Looks like PGV rule is not actually implemented here. I think it is safer to have this then before PGV is implemented

File: source/extensions/filters/http/composite/action.cc
Comment (tyxia) [reply source/extensions/filters/http/composite/action.cc:119]: Thanks for explanation! coin toss in data plane makes sense.  It was a bit implicit under callback. But please see my another comment https://github.com/envoyproxy/envoy/pull/34611#issuecomment-2173883851 for further discussion

**Comment by yanjunxiang-google** [comment]
> @yanjunxiang-google After closer look at this PR, I am think if we should move sampling to `extensionWithMatcher`/match_delegate filter:
> 
> (1) It will be more natural. The sampling in the PR is actually performed in match_delegate filter, just in a bit implicit callback format. Doing it in match_delegate with its own sampling config seems to be more natural.
> 
> (2) It will be more generic. Essentially, composite action is just one of `TypedExtensionConfig` in extensionWithMatcher. Doing it in match_delegate opens up future opportunity, especially given (1) that sampling is already done there.
> 
> (3) It also doesn't introduce changes/complexity to xDS matcher and limit the scope of the change to extension/filter, like this PR (address API concern)
> 
> WDYT?

If your idea is something like directly adding a sampling_ratio API under the matcher, then I am very hesitate to do that, as the current xDS matcher API is already very complicated.

I am supporting adding it under a `TypedExtensionConfig` to be away with the matcher proto.  I had been thinking adding another `TypedExtensionConfig` action' with sampling and actual filter, then at the end that will be almost identical to the existing ExecuteFilterAction with sampling, that's the reason I went ahead with this approach.

**Comment by tyxia** [comment]
> > @yanjunxiang-google After closer look at this PR, I am think if we should move sampling to `extensionWithMatcher`/match_delegate filter:
> > (1) It will be more natural. The sampling in the PR is actually performed in match_delegate filter, just in a bit implicit callback format. Doing it in match_delegate with its own sampling config seems to be more natural.
> > (2) It will be more generic. Essentially, composite action is just one of `TypedExtensionConfig` in extensionWithMatcher. Doing it in match_delegate opens up future opportunity, especially given (1) that sampling is already done there.
> > (3) It also doesn't introduce changes/complexity to xDS matcher and limit the scope of the change to extension/filter, like this PR (address API concern)
> > WDYT?
> 
> If your idea is something like directly adding a sampling_ratio API under the matcher, then I am very hesitate to do that, as the current xDS matcher API is already very complicated and hard to understand.
> 
> I am supporting adding it under a `TypedExtensionConfig` to be away with the matcher proto. I had been thinking adding another `TypedExtensionConfig` action with sampling and actual filter, then at the end that will be almost identical to the existing ExecuteFilterAction with sampling, that's the reason I went ahead with this approach.

I am suggesting `message ExtensionWithMatcher` which is filter config like composite filter. Also, this doesn't touch xDS matcher

File: source/extensions/filters/http/match_delegate/config.cc
Comment (tyxia) [inline source/extensions/filters/http/match_delegate/config.cc:123]: @yanjunxiang-google , I think the sampling is actually performed at `match_result.result_()` as callback, right?

My comment above are :
(1) Since the sampling is already performed in match_delegate filter, It seems to be more natural to have the sampling config in its own filter config?
(2) It is more generic as other `TypedExtensionConfig` besides composite action can also make use of sampling feature under match_delegate.

Comment (tyxia) [reply source/extensions/filters/http/match_delegate/config.cc:123]: cc @htuch  and @rshriram for second opinion

Comment (yanjunxiang-google) [reply source/extensions/filters/http/match_delegate/config.cc:123]: we can also interpret the current approach has done sampling inside the action callbacks, and match_delegate filter  is not aware of sampling or whatsoever, and it does not have to know.

**Comment by yanjunxiang-google** [comment]
> > > @yanjunxiang-google After closer look at this PR, I am think if we should move sampling to `extensionWithMatcher`/match_delegate filter:
> > > (1) It will be more natural. The sampling in the PR is actually performed in match_delegate filter, just in a bit implicit callback format. Doing it in match_delegate with its own sampling config seems to be more natural.
> > > (2) It will be more generic. Essentially, composite action is just one of `TypedExtensionConfig` in extensionWithMatcher. Doing it in match_delegate opens up future opportunity, especially given (1) that sampling is already done there.
> > > (3) It also doesn't introduce changes/complexity to xDS matcher and limit the scope of the change to extension/filter, like this PR (address API concern)
> > > WDYT?
> > 
> > 
> > If your idea is something like directly adding a sampling_ratio API under the matcher, then I am very hesitate to do that, as the current xDS matcher API is already very complicated and hard to understand.
> > I am supporting adding it under a `TypedExtensionConfig` to be away with the matcher proto. I had been thinking adding another `TypedExtensionConfig` action with sampling and actual filter, then at the end that will be almost identical to the existing ExecuteFilterAction with sampling, that's the reason I went ahead with this approach.
> 
> I am suggesting `message ExtensionWithMatcher` which is match_delegate filter config similar to composite filter here. Also, this doesn't touch xDS matcher

File: source/extensions/filters/http/match_delegate/config.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/match_delegate/config.cc:123]: Also, I feel sampling fits well for the composite filter logic:  if sampling returns positive, then take the action, otherwise, skip.  So, I guess the existing solution resolves both https://github.com/envoyproxy/envoy/issues/30147 and https://github.com/envoyproxy/envoy/issues/30006.

Another thing is that adding the sampling under the wrapper of xDS matcher requires to add the sample_percent API in both ExtensionWithMatcher and ExtensionWithMatcherPerRouter.  From API perspective, not sure whether this is clean or not.

Comment (tyxia) [reply source/extensions/filters/http/match_delegate/config.cc:123]: But the current approach is not generic because it only limits to composite action callback. If other actions need sampling, you will need to either duplicate sampling or force them to use composite action.

Also, as change here shows, match_delegate must invoke the callback and check the result to make sampling work. So it is not transparent to match_delegate .

Comment (yanjunxiang-google) [reply source/extensions/filters/http/match_delegate/config.cc:123]: Yeah, I think ask customer to use composite action probably is okay for the existing issues https://github.com/envoyproxy/envoy/issues/30147 and https://github.com/envoyproxy/envoy/issues/30006.  As stated above, composite filter has the characteristics either executing the action or skipping is a good fit for the sampling behavior. 

For the change here, adding a result nullptr check is intuitive to me before call result->typeUrl(). We should add this nullptr check regardless whether the underline filter callback has sampling or not.

There is another issue if putting the sampling under the ExtensionWithMatcher but not in composite action:  Let's say you have a matcher list with three matchers, and one matcher to ext_proc, which you want to do sampling with 10% ratio; Another matcher to RBAC filter, which you want to do sampling with 100% ratio; The third matcher with access_logging, you want to do sampling with 1%.  Putting sampling under ExtensionWithMatcher wont' support such kind of flexibility as you only have one ratio for all underline matchers. But putting it under action can.

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (mattklein123) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:67]: I would use `core.v3.RuntimeFractionalPercent` so this can be controlled via Runtime as well if desired.

**Review by mattklein123 — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/match_delegate/config.cc
Comment (tyxia) [reply source/extensions/filters/http/match_delegate/config.cc:123]: I think `ExtensionWithMatcher` can still support the example above (which I agree it's a legit use case) with a map of sampling ratios. Action name will be the key to look up ratio for matched action filter and sampling can be performed accordingly. This will introduce certain implementation complexity to match_delegate though. 
With that said, I am ok with current approach to unblock our existing needs. In the long term, we probably should think about xDS matcher approach to make it ultimately flexible.

**Review by tyxia — DISMISSED** [review]
LGTM. Thanks!

Please address Matt's comment

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto]: Good idea, done!

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:62]:
```suggestion
  // Probability of the action execution. If not specified, this is 100%.
```

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:63]:
```suggestion
  // This allows sampling behavior for configured actions.
```

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:67]:
```suggestion
  // probability is done. The underline action will only be executed if the
```

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:68]:
```suggestion
  // dice roll returns positive. Otherwise, the action is skipped.
```

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/composite/v3/composite.proto:69]: You do not need this comment as it is available in the config.core.v3.RuntimeFractionalPercent documentation

File: changelogs/current.yaml
Comment (yanavlasov) [inline changelogs/current.yaml:39]:
```suggestion
    It specifies probability of the action execution. If not specified, it is 100%.
```

File: source/extensions/filters/http/composite/action.cc
Comment (yanavlasov) [inline source/extensions/filters/http/composite/action.cc:152]: The `isSampled` function is called on the dataplane threads. As such it can not throw C++ exceptions. But the getSampleRatio that is called by isSampled throws exceptions.

Also note that `getSampleRatio` parses the configuration on every sample check. Config parsing and error checking should be done once and only runtime value is queried on every sample.

File: source/extensions/filters/http/composite/action.cc
Comment (yanavlasov) [inline source/extensions/filters/http/composite/action.cc:34]: Exceptions are currently being removed as error reporting mechanism. To make this process easier please use `absl::StatusOr<float>` and only throw at the top level factory method.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto]: done

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto]: done

File: api/envoy/extensions/filters/http/composite/v3/composite.proto
Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/composite/v3/composite.proto]: done

File: changelogs/current.yaml
Comment (yanjunxiang-google) [reply changelogs/current.yaml]: Good idea. done

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc]: Good catch.  Separated the configuration parsing (getSampleRatio()), and the runtime key retrieving code(getSampleRatioRuntime()). getSampleRatio() is now called in the control plane configuration parsing phase. And isSampled->getSampleRatioRuntime() is called in data plane.

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc:33]: As this function is called in a couple of places. To avoid throw the exceptions in multiple places, I prefer just throw it here.

File: source/extensions/filters/http/composite/action.cc
Comment (mattklein123) [inline source/extensions/filters/http/composite/action.cc:44]: I'm very surprised this type of error checking doesn't already exist elsewhere in the code base. If it doesn't can you move it to where the other runtime helpers live?

File: source/extensions/filters/http/composite/action.cc
Comment (mattklein123) [inline source/extensions/filters/http/composite/action.cc:62]: There are other helpers in runtime that already do this stuff for fractional percent. See https://github.com/envoyproxy/envoy/blob/main/source/common/runtime/runtime_impl.cc.

**Review by mattklein123 — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc]: Good idea!  Changed to use existing runtime.snapshot().featureEnabled() API for the sampling check.

File: source/extensions/filters/http/composite/action.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/composite/action.cc]: Done

**Comment by yanjunxiang-google** [comment]
Kind Ping!

**Comment by yanjunxiang-google** [comment]
Kind Ping!

**Review by htuch — COMMENTED** [review]
/lgtm api

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
Waiting for @mattklein123 LGTM

**Comment by yanjunxiang-google** [comment]
Great! Thanks everyone for reviewing the PR and approval!

**Pull request merged**
